openfisca_core/data_storage.py

```diff
@@ -1,136 +1,191 @@
-# -*- coding: utf-8 -*-
-
-import shutil
+import abc
 import os
+import shutil
+from typing import Dict, List, Optional
 
-import numpy as np
+import numpy
 
+from openfisca_core import indexed_enums
 from openfisca_core import periods
-from openfisca_core.periods import ETERNITY
-from openfisca_core.indexed_enums import EnumArray
 
+from typing_extensions import Protocol
+
+
+class StorageLike(Protocol):
```

Comment:
Mixing abstract base classes and structural subtyping feels weird here.

Reply:
Why does it feel weird ? In this case, as I'm not actually implementing anything, I could effectively just pick structural subtyping at not cost.

https://www.python.org/dev/peps/pep-0544/#explicitly-declaring-implementation

Reply:
Picked base classes in the end.